path/path.go

```diff
@@ -3,188 +3,301 @@ package path
-			return "", &ErrInvalidPath{error: err, path: txt}
+	// If the path doesn't begin with a "/", we expect it to start with a CID and
+	// be an IPFS Path.
+	if components[0] != "" {
+		root, err := cid.Decode(components[0])
+		if err != nil {
+			return nil, &ErrInvalidPath{error: err, path: str}
 		}
-		// The case when the path starts with hash without a protocol prefix
-		return Path("/ipfs/" + txt), nil
+
+		return &path{
+			str:       cleaned,
+			root:      root,
+			namespace: IPFSNamespace,
+		}, nil
 	}
 
-	if len(parts) < 3 {
-		return "", &ErrInvalidPath{error: fmt.Errorf("invalid ipfs path"), path: txt}
+	if len(components) < 3 {
+		return nil, &ErrInvalidPath{error: fmt.Errorf("not enough path components"), path: str}
 	}
 
-	//TODO: make this smarter
-	switch parts[1] {
-	case "ipfs", "ipld":
-		if parts[2] == "" {
-			return "", &ErrInvalidPath{error: fmt.Errorf("not enough path components"), path: txt}
+	switch components[1] {
+	case IPFSNamespace, IPLDNamespace:
+		if components[2] == "" {
+			return nil, &ErrInvalidPath{error: fmt.Errorf("not enough path components"), path: str}
 		}
-		// Validate Cid.
-		_, err := decodeCid(parts[2])
+
+		root, err := cid.Decode(components[2])
 		if err != nil {
-			return "", &ErrInvalidPath{error: fmt.Errorf("invalid CID: %w", err), path: txt}
+			return nil, &ErrInvalidPath{error: fmt.Errorf("invalid CID: %w", err), path: str}
 		}
-	case "ipns":
-		if parts[2] == "" {
-			return "", &ErrInvalidPath{error: fmt.Errorf("not enough path components"), path: txt}
+
+		return &path{
+			str:       cleaned,
+			root:      root,
+			namespace: components[1],
+		}, nil
+	case IPNSNamespace:
+		if components[2] == "" {
+			return nil, &ErrInvalidPath{error: fmt.Errorf("not enough path components"), path: str}
 		}
-	default:
-		return "", &ErrInvalidPath{error: fmt.Errorf("unknown namespace %q", parts[1]), path: txt}
-	}
 
-	return Path(txt), nil
-}
+		var root cid.Cid
+		pid, err := peer.Decode(components[2])
+		if err != nil {
+			// DNSLink.
+			root = cid.Undef
+		} else {
+			root = peer.ToCid(pid)
+		}
 
-// ParseCidToPath takes a CID in string form and returns a valid ipfs Path.
-func ParseCidToPath(txt string) (Path, error) {
-	if txt == "" {
-		return "", &ErrInvalidPath{error: fmt.Errorf("empty"), path: txt}
+		return &path{
+			str:       cleaned,
+			root:      root,
+			namespace: IPNSNamespace,
+		}, nil
+	default:
+		return nil, &ErrInvalidPath{error: fmt.Errorf("unknown namespace %q", components[1]), path: str}
 	}
+}
 
-	c, err := decodeCid(txt)
-	if err != nil {
-		return "", &ErrInvalidPath{error: err, path: txt}
+// NewPathFromSegments creates a new [Path] from the provided segments. This
+// function simply calls [NewPath] internally with the segments concatenated
+// using a forward slash "/" as separator.
+func NewPathFromSegments(segments ...string) (Path, error) {
+	if len(segments) > 1 {
+		if segments[0] == "" {
+			segments = segments[1:]
+		}
 	}
```

Comment:
Why are we doing this? It seems like either we're relying on `NewPath` to do things like `gopath.Clean` and erroring on paths without enough segments, or we're not. Will removing this result in problems?

Reply:
Honestly, I'm not sure why I added this. I removed it. I have to add more tests.